Support external config schemas in the backend

davidfestal opened the pull request:
## Hey, I just made a Pull Request!

On the backend application, thisPR allows referencing additional configuration schemas at runtime (application start), and have them taken in account where schemas are used:
- the backed-app plugin, in order to prepare the frontend configuration
- the backend application loggers, in oder to hide secret configuration values (defined in the schemas)

This fixes #21631

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

davidfestal commented:
@Rugvip Could this PR be reopened ? Thanks !

freben commented:
Sure!

davidfestal commented:
@Rugvip @freben Could this PR be reopened ? Thanks !
Any idea whether it could get a review ?

davidfestal commented:
@Rugvip @backstage-service Are these errors in the e2e tests known: https://github.com/backstage/backstage/actions/runs/7463481178/job/20308264561?pr=21632#step:11:143

Or do they ring a bell to anyone ?

This seems to be something frontend-related, and I don't see how it would have anyting to do with the changes done in this PR (at least afaict). Any help would be appreciated.

davidfestal commented:
@Rugvip This has been rebased and completed.

Rugvip commented on packages/backend-plugin-api/src/alpha.ts:
```diff
@@ -33,3 +36,32 @@ export const featureDiscoveryServiceRef =
+
+/** @alpha */
+export interface SchemaDiscoveryService {
+  getAdditionalSchemas(): Promise<{ schemas: Array<ConfigSchemaPackageEntry> }>;
```
Do you think it would make sense to lift the abstraction up a level or two here? Instead of providing additional schemas it could instead have have this either provide the entire schema, or even the entire schema processing interface (as in `ConfigSchema['process']`. It does make the app schema loading more tricky, but then again we might need some different solution for that anyway as discussed separately.

davidfestal replied:
I'm not convinced at all it would be worth the refactoring cost.

davidfestal replied:
Doing it this way makes the addition of an external config schema provider very easy, without having to interfere with the schema loading implementation, which is a core thing.

davidfestal replied:
If you think it is needed to have this PR merged, I can even isolate the implementation of this new `alpha` service, provided in the `backend-dynamic-feature-service` package, in a dedicated `alpha` API.

Rugvip commented on packages/config-loader/api-report.md:
```diff
@@ -45,6 +45,12 @@ export type ConfigSchema = {
 };
 
+// @public
+export type ConfigSchemaPackageEntry = {
+  value: JsonObject;
+  path: string;
+};
```
The `path` here is really just there for more informative error messages when validating the schema. I think it'd be best to avoid exporting this type and then do validation upfront instead, although that might be tricky. Alternatively it would likely make sense to rename this to `context` or something like that instead.

Similarly, `value` isn't a particularly good name here imo, probably best to have that be `schema`, `jsonSchema`, or something like that instead.

davidfestal replied:
Should I then rename both `path` to `context` and `value` to `jsonSchema` and still export it ?

davidfestal replied:
Done in commit e896620f2c1ca8d0feff24a6cf0d0c880bd6809e

davidfestal replied:
Well, this breaks unit tests for a number of plugins: https://github.com/backstage/backstage/actions/runs/7610309427/job/20723424247?pr=21632

That's because they implicitly depend on the field names `path` and `value` when testing again expected serialized JSON Values.

Should I revert this change ?
Or create a dedicated type with `context` and `jsonSchema` fields to be used in this PR ?

davidfestal replied:
In addition to that, I find other places where the `value` field is referenced like: https://github.com/backstage/backstage/blob/master/packages/cli/src/commands/config/schema.ts#L35.

This makes me think that I should not rename those fields now.

davidfestal replied:
For now I reverted the commit mentioned above and kept the existing names in the exported type, knowing that the field names are expected to be known in a number of places throughout the codebase.

davidfestal replied:
In fact I just made the `ConfigSchemapackageEntry` non public anymore, and simply used a map of JsonObject indexed on strings (the context of the external schema, most likely a path).
That seems clearer and doesn't impact the existing naming of the existing fields `ConfigSchemapackageEntry`, no exposes it.

The change is done in commit 84fcbc49d6d444bcfb347886ed8ca33cb139954b

Rugvip commented on plugins/app-backend/src/lib/config.ts:
```diff
@@ -87,7 +89,10 @@ export async function readConfigs(options: ReadOptions): Promise<AppConfig[]> {
 
     try {
-      const schema = await loadConfigSchema({ serialized: serializedSchema });
+      const schema = await loadConfigSchema({
+        serialized: serializedSchema,
+        additionalSchemas: options.additionalSchemas,
+      });
```
Wanna note here that I think we'll move away from the need for frontend config schema for plugins. The new frontend system basically has all config under `app.*` and I think we may even be able to get rid of the `ConfigApi` completely in the frontend. Basically we hopefully won't need to bother with this in the future.

davidfestal replied:
:+1:

Rugvip replied:
Had another thing about what we could do here. I'm thinking the best design here might be to introduce an extension point for the app backend, along with an option for the router. I think the best case again would be if we could make that something that provides the whole `ConfigSchema` rather than additional entries. Seems like that shouldn't be too bad? bit of duplicated code for loading the serialized schema perhaps but I think that prolly worth it.

Rugvip commented on packages/backend-dynamic-feature-service/src/scanner/plugin-scanner.ts:
```diff
@@ -313,3 +321,70 @@ export class PluginScanner {
+   * @param platform - The platform of the plugin.
+   * @returns the plugin-relative path to the Json schema file.
+   */
+  schemaLocator?: (platform: PackagePlatform) => string;
```
`PackagePlatform` is just `node`, `web` or `common`. I think this could be something like `schemaLocationByPlatform: { [platform in PackagePlatform]: path }` instead?

davidfestal replied:
> I think this could be something like schemaLocationByPlatform: { [platform in PackagePlatform]: path } instead

Wouldn't doing this force us to duplicate the value when it is the same value for both cases ?

As an example the `schemaLocator` argument default value in `gatherDynamicPlginsSchemas()` would change from:

```ts
  schemaLocator: (platform: PackagePlatform) => string = () =>
    path.join('dist', 'configSchema.json'),
```
to
```ts
  schemaLocationByPlatform: { [platform in PackagePlatform]: string } = {
    'web': path.join('dist', 'configSchema.json'),
    'node': path.join('dist', 'configSchema.json'),
  }
```
I can do the change if you prefer, but I'm not convinced it is better.
I'd rather give even more freedom to the caller and pass the whole `ScannedPluginPackage` as an argument.

Rugvip replied:
Yeah it was more a comment on the callback not being very powerful tbh. I think it make sense to provide more information instead, and keep the callback.

davidfestal replied:
Let me change it.

davidfestal replied:
Done in commit a9e95d6ad1770745c6493de87965d97bf1090ccd

Rugvip commented on .changeset/swift-pumpkins-shake.md:
```diff
@@ -0,0 +1,12 @@
+---
+'@backstage/backend-app-api': minor
+'@backstage/backend-common': minor
+'@backstage/config-loader': minor
+'@backstage/plugin-app-backend': minor
+'@backstage/backend-dynamic-feature-service': minor
```
Could you split this up in a couple more changesets that more accurately describe the changes to each package?

davidfestal replied:
Done in commit 7f74e6409bf781eb248c8392cf19e7f58a90aab9

Rugvip commented on packages/backend-dynamic-feature-service/src/scanner/schemas.ts:
```diff
@@ -0,0 +1,75 @@
+  const allSchemas: { value: any; path: string }[] = [];
+
+  for (const pluginPackage of packages) {
+    const platform = PackageRoles.getRoleInfo(
```
I find this to be a bit tricky for frontend plugins, since it relies on them actually being installed in the `app-backend` instance. In the long run that's not gonna be the case right? There might be a couple of plugins brought in from a separate endpoint. Not really something to address as part of this PR, it's more of just another reason for why I want to move away from using config outside of `app.` for the frontend.

davidfestal replied:
The dynamic plugin scanner scans for, and discovers, both backend and frontend plugins (even though the loading mechanism is different for frontend and backend).
And since config schemas are still used for frontend plugins, we have to handle both cases here.

> Not really something to address as part of this PR, it's more of just another reason for why I want to move away from using config outside of app. for the frontend.

I get it. :+1:

Rugvip reviewed (changes requested):
Nice! Glad to see the design with a separate experimental service 👌 👍 

Couple of API concerns that need to be addressed, but also a bunch of higher level stuff that doesn't necessarily need to be addressed as part of this PR.

davidfestal commented:
@Rugvip Is the PR now OK for you, or do you want some other option to be implemented for comment https://github.com/backstage/backstage/pull/21632#discussion_r1461001122 ?

davidfestal commented:
> @Rugvip Is the PR now OK for you, or do you want some other option to be implemented for comment [#21632 (comment)](https://github.com/backstage/backstage/pull/21632#discussion_r1461001122) ?

@Rugvip Could you please remove the `stale` label ? It seems to me that this PR was quite near to be merged, though I was still waiting for answer to my question above. This PR is quite important to have the dynamic pkugins be really usable in upstream backstage.
Do you think you could provide me with your thoughts about what you expect me to further fix to have it merged ?

davidfestal commented:
@Rugvip Please see comment https://github.com/backstage/backstage/pull/21632#discussion_r1480368731

It seems that now all your comments / requests for change have been fixed.
I have also rebased the PR on top of master.

Is it OK to merge it now ?

cc @christophe-f

davidfestal commented:
> @Rugvip Please see comment [#21632 (comment)](https://github.com/backstage/backstage/pull/21632#discussion_r1480368731)
> 
> It seems that now all your comments / requests for change have been fixed. I have also rebased the PR on top of master.
> 
> Is it OK to merge it now ?
> 
> cc @christophe-f

In the following comment I added a more complete answer to your question about lifting up the API one level or two : https://github.com/backstage/backstage/pull/21632#discussion_r1482778993

Hopefully it would help.

Rugvip commented on packages/backend-app-api/src/config/config.ts:
```diff
@@ -29,16 +29,19 @@ import type { Config, AppConfig } from '@backstage/config';
 }): Promise<(config: Config) => Iterable<string>> {
   const { logger, dir = process.cwd() } = options;
   const { packages } = await getPackages(dir);
   const schema = await loadConfigSchema({
     dependencies: packages.map(p => p.packageJson.name),
+    additionalSchemas: options.additionalSchemas,
   });
```
This is a key place where I think we could introduce an abstraction that should make you able to set this up as needed. I'm thinking we could add an optional `schema` option to `createConfigSecretEnumerator`, `schema?: ConfigSchema`? If a schema is provided it simply won't to try to load it itself.

On top of that I think the `packagePaths` option for `loadConfigSchema` might actually be enough for you load load in additional schemas btw? But if it's not we can add the `additionalSchemas` option or similar.

davidfestal replied:
Done.
I even removed the `additionalSchemas` option.

Rugvip commented on packages/backend-common/api-report.md:
```diff
@@ -560,6 +561,9 @@ export function loadBackendConfig(options: {
   logger: LoggerService;
   remote?: LoadConfigOptionsRemote;
   additionalConfigs?: AppConfig[];
+  additionalSchemas?: {
+    [context: string]: JsonObject;
+  };
```
No need to add this right? or are you relying on the old backend system? Either way it'd be better to move to the new APIs for loading config: https://github.com/backstage/backstage/blob/fb50191fa3a289a990e4de1098525da5d6c9f6ae/packages/backend-app-api/src/services/implementations/config/rootConfigServiceFactory.ts#L45-L50

davidfestal replied:
> are you relying on the old backend system

No, the whole point of this PR is to enable us relying on the new backend system.

So you're right, let's remove it.

Rugvip reviewed (changes requested):
Alright came back to this now with a fresh set of eyes and I might've found some ways forward that untangle things a bit. Hoping you find it to be a good solution both for now and for a little longer in the future. I do agree that lifting the abstraction level even further might trigger a fairly sizeable refactor and I don't mind avoiding that for now.

One of the things I actually think is best to avoid is to have the two backend schema and frontend schema usages be too tightly connected. They're quite different after all, and in practice they generally use completely different sets of schemas, so to then have a single service that feed both I think would be unfortunate. That's especially with the goal of eventually removing the need for frontend schema at some point altogether.

davidfestal commented:
@Rugvip I applied your proposal in commit https://github.com/backstage/backstage/pull/21632/commits/2fb527dd34dab57917005d6e0b51a1c5332d4f99 so that:
- there is no more alpha `schemaDiscovery` service
- minimal changes to the API boil down to providing an option to pass an optional `ConfigSchema` to either the  `createConfigSecretEnumerator` or the `app-backend` plugin (through an extension)
- Dynamic plugin schema support is added by an alternate `rootLogger` service implementation and an `app-backend` plugin module available in the `backend-dynamic-features-service`. 

Hopefully this design would make this PR mergeable ?

davidfestal commented:
> @Rugvip I applied your proposal in commit [2fb527d](https://github.com/backstage/backstage/commit/2fb527dd34dab57917005d6e0b51a1c5332d4f99) so that:
> 
> * there is no more alpha `schemaDiscovery` service
> * minimal changes to the API boil down to providing an option to pass an optional `ConfigSchema` to either the  `createConfigSecretEnumerator` or the `app-backend` plugin (through an extension)
> * Dynamic plugin schema support is added by an alternate `rootLogger` service implementation and an `app-backend` plugin module available in the `backend-dynamic-features-service`.
> 
> Hopefully this design would make this PR mergeable ?

However the verify check fails due to [this error]( https://github.com/backstage/backstage/actions/runs/7876518192/job/21490774805?pr=21632#step:9:10):

> Plugin package '@backstage/backend-dynamic-feature-service' with role 'node-library' has a runtime dependency on package '@backstage/backend-app-api', which is not permitted. If you are using this dependency for dev server purposes, you can move it to devDependencies instead.

Should the `createConfigSecretEnumerator()` function be moved to another package ?

Rugvip reviewed (approved):
Nice, looks good! 👍 🎉 

Let's :shipit:

Rugvip commented on .changeset/late-poets-love.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage/backend-app-api': patch
+---
+
+Allow the `createConfigSecretEnumerator` to take an optional `configSchema` argument with an already-loaded global configuration schema.
```
```suggestion
Allow the `createConfigSecretEnumerator` to take an optional `schema` argument with an already-loaded global configuration schema.
```

Rugvip commented on .changeset/rich-bees-fry.md:
```diff
@@ -0,0 +1,6 @@
+---
+'@backstage/plugin-app-backend': minor
+'@backstage/plugin-app-node': minor
```
```suggestion
'@backstage/plugin-app-backend': patch
'@backstage/plugin-app-node': patch
```

No breaking changes here

Rugvip commented:
> Should the createConfigSecretEnumerator() function be moved to another package ?

Yep we've been discussing this a bit, and we may move things out of `backend-app-api`. Current solution is good for now 👍
